sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -40,28 +52,38 @@ void ScheduleCleanup()
                 }
 
                 this.cleanupScheduled = true;
-                Task.Run(async () => await this.CollectExpiredEntriesAsync().ConfigureAwait(false));
+                _ = this.CollectExpiredEntriesAsync(token);
             }
         }
 
-        async Task CollectExpiredEntriesAsync()
+        async Task CollectExpiredEntriesAsync(CancellationToken token)
         {
-            await Task.Delay(delayBetweenCleanups);
-
-            lock (this.cleanupSynObject)
+            try
             {
-                this.cleanupScheduled = false;
+                await Task.Delay(delayBetweenCleanups, token).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                return;
+            }
+            finally
+            {
+                lock (this.cleanupSynObject)
+                {
+                    this.cleanupScheduled = false;
+                }
             }
 
-            foreach (var key in this.dictionary.Keys)
+            foreach (var kvp in this.dictionary)
             {
-                if (DateTime.UtcNow > this.dictionary[key])
+                var expiration = kvp.Value;
+                if (DateTime.UtcNow > expiration)
                 {
-                    this.dictionary.TryRemove(key, out _);
+                    this.dictionaryAsCollection.Remove(kvp);
```

Comment:
I believe this will modify the collection while you are iterating over it, won't it?

Since both `dictionary` and `dictionaryAsCollection` point to the same object.

Reply:
See https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getenumerator